src/execution_plan/execution_plan.c

```diff
@@ -248,40 +248,83 @@ ExecutionPlan* NewExecutionPlan(RedisModuleCtx *ctx, Graph *g, const char *graph
         execution_plan->filter_tree = BuildFiltersTree(ast->whereNode->filters);
 
     if(ast->matchNode) {
-        BuildQueryGraph(ctx, g, graph_name, q, ast->matchNode->graphEntities);
-
-        /* Do we required to perform traversal? */
-        if(q->edge_count == 0) {
-            /* Node scan. */
-            Node *n = q->nodes[0];
-            if(n->label)
-                op = NewOpNode(NewNodeByLabelScanOp(ctx, q, g, graph_name, q->nodes, n->label));
-            else
-                op = NewOpNode(NewAllNodeScanOp(q, g, q->nodes));
-            Vector_Push(ops, op);
-        } else {
-            size_t expCount = 0;
-            AlgebraicExpression **exps = AlgebraicExpression_From_Query(ast, q, &expCount);
+        BuildQueryGraph(ctx, g, graph_name, q, ast->matchNode->_mergedPatterns);
+
+        // For every pattern in match clause.
+        size_t patternCount = Vector_Size(ast->matchNode->patterns);
+        
+        /* Incase we're dealing with multiple patterns
+         * we'll simply join them all together with a join operation. */
+        bool multiPattern = patternCount > 1;
+        OpNode *cartesianProduct = NULL;
+        if(multiPattern) {
+            cartesianProduct = NewOpNode(NewCartesianProductOp());
```

Comment:
There's no way around it, of course, but I'm scared just thinking of the cost of a Cartesian product with _n_ operands! Ick.